### ISSUE TITLE
This library is stable for production? (expo & bare app)

### ISSUE THREAD

**Issue by maoryadin** [twilio/twilio-voice-react-native#132]
This library is stable for production? i am wondering if it would work with expo/react native bare app.

**Comment by mhuynh5757** [comment]
Hi @maoryadin, thanks for reaching out! The library is stable and we're hard at work on a beta release, we don't expect the API to change in major ways. We haven't tested the SDK in Expo applications, but we have seen some of our customers getting it working on Expo. See related GitHub issues:
https://github.com/twilio/twilio-voice-react-native/issues/102
https://github.com/twilio/twilio-voice-react-native/issues/117
https://github.com/twilio/twilio-voice-react-native/issues/128

**Comment by maoryadin** [comment]
Adding native code to Expo can be challenging, as it requires some expertise.

```
@interface AppDelegate ()

@property (nonatomic, strong) TwilioVoicePushRegistry *twilioVoicePushRegistry;

@end

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // ...

    // Initialize PKPushRegistry at launch
    self.twilioVoicePushRegistry = [TwilioVoicePushRegistry new];
    [self.twilioVoicePushRegistry updatePushRegistry];

    return YES;
}
```

Do you have plans to offer support or develop an Expo module for this functionality?
It also necessary to add PushKit for iOS, which can be challenging to achieve with Expo.

**Comment by charliesantos** [comment]
Hi @maoryadin , we're currently evaluating whether we will support it, and if so, when. Feedback like yours definitely helps on prioritization. Let's keep this issue open and get a feel of the customer demand.

**Comment by ovidb** [comment]
We would also be interested in support for Expo. I'm sure you're already familiar but probably the easiest way to go at it is to create a [expo config plugin](https://docs.expo.dev/guides/config-plugins/)

**Comment by mhuynh5757** [comment]
Hi @ovidb, thanks for the feedback. We will take this into consideration when we align priorities as a team. It's on our radar!

**Comment by dwbelliston** [comment]
Does anyone have a POC of working with expo? Ive not had any luck yet, but also very unfamiliar with ios/android configs.

The root error im getting after adding to an expo project in expo go is that javascript is trying to access native code. Which makes sense since react-native-twilio is using native modules not provided by the expo go client.

So, im doing a custom dev client (still working on that), but I am not sure which native modules I am need to add?

Which file in the ios folder shows the modules being used? 

Any tips would be awesome!

**Comment by maoryadin** [comment]
> Does anyone have a POC of working with expo? Ive not had any luck yet, but also very unfamiliar with ios/android configs.
> 
> The root error im getting after adding to an expo project in expo go is that javascript is trying to access native code. Which makes sense since react-native-twilio is using native modules not provided by the expo go client.
> 
> So, im doing a custom dev client (still working on that), but I am not sure which native modules I am need to add?
> 
> Which file in the ios folder shows the modules being used?
> 
> Any tips would be awesome!

 Yes, I was able to run the package with Expo, I will share the process when I get on my computer (probably on Sunday).

**Comment by dwbelliston** [comment]
I was able to make a bit more progress.

For others that might be trying to figure out the managed Expo path here is my journey so far:

After looking at this: https://github.com/twilio/twilio-voice-react-native/pull/143

I realized that this project itself is doing the pushkit registration, so that is not something I need to try to add in manually. 

So, what I did:
- From my expo project, i added the @twilio/voice-react-native-sdk package through yarn.
- Then, you need to create the 'custom dev client' through the expo, they have good docs on that but you run something like
`eas build --profile development --platform ios`
- After that build works, you can install on your personal real device
- After that client is downloaded to your device, you should see your app icon
- The start the dev client server from your computer
`npx expo start --dev-client`
- Opening up the app icon, I was able to run the app on my dev client, and make a call!

I need to setup the call handling and the incoming call notification still, but excited about being able to do an outbound call, without needed to 'eject' from expo

To all those building the project, thanks for the effort on this!

**Comment by MykolaKobetiakCloudGenie** [comment]
@dwbelliston Hi there, were you able to set up incoming calls? I have trouble with them. Outgoing works fine, but I am not receiving Voice.Event.CallInvite on IOS and Android.

**Comment by dwbelliston** [comment]
Hey @MykolaKobetiakCloudGenie - sorry for the delay, I didnt make any more progress on this.

A while back I reached out to the Twilio team and they mentioned they are going to be adding some more examples of how to handle incoming calls and so I decided to wait until those were available.

Not sure the status of it now

**Comment by MykolaKobetiakCloudGenie** [comment]
@dwbelliston I did prebuild, and additional push notifications Java handler. Now it works. This info is helpful: #160

**Comment by mhuynh5757** [comment]
Hi all, thanks for the vibrant community discussion. As @dwbelliston mentioned, we were working on more examples for how to handle outgoing and incoming calls. Please see the Reference App for more details.

https://github.com/twilio/twilio-voice-react-native-app

**Comment by amitava82** [comment]
An example app with expo would be very helpful.

**Comment by dwbelliston** [comment]
Hey! Looks like this app has been making steady progress. Would one of the owners be willing to give a little update on stability/roadmap? Really eager to start building on this, but hesitant do to the beta tag. Seriously though, thanks for the effort!

**Comment by hawkinbj** [comment]
@dwbelliston PM here, we're preparing to remove the Beta tag and release to GA in the next week or so (very close!). This will be a first class SDK flavor in the Twilio Voice product suite. It's production ready as-is as well so hopefully that removes any concerns around deciding whether or not to start building on it. We'd love to hear your feedback on how to improve for future roadmap considerations as you and team get in there and start getting your hands dirty!

**Comment by MykolaKobetiakCloudGenie** [comment]
Hi @hawkinbj  , thank you for the updates. Will this issue #341 be fixed in the next release you mentioned?

**Comment by hawkinbj** [comment]
@MykolaKobetiakCloudGenie yes it will be, it's currently baked into the GA release that is coming shortly

**Comment by amitava82** [comment]
@hawkinbj This library isn't usable without resolving https://github.com/twilio/twilio-voice-react-native/issues/295
I've spent several hours and still clueless how to use this library.

**Comment by dwbelliston** [comment]
So excited to use it! Just in time for our roadmap priorities too. Will dive in and provide features, if you want/need a use case for a niche tech startup that is using it to offer compliance text/calls to financial advisors happy to talk.

**Comment by mhuynh5757** [comment]
Hey @amitava82 thanks for bringing up your concerns about our documentation. We're always looking for ways to help onboard and make our product as easy to use. We have a migration guide that's performing double duty as supplementary Android documentation:

https://github.com/twilio/twilio-voice-react-native/blob/main/docs/migration-guide-beta.4.md

But I understand it doesn't have very good visibility. As a customer, would you like to see this replaced or made more visible (i.e. through our readme), or is this document insufficient in your opinion? What other changes would you like to see in our docs?

**Comment by mhuynh5757** [comment]
Just FYI @amitava82 we saw your comment in #295 and we acknowledge and understand. Let's see what we can do to improve the user experience for everyone.

**Comment by mhuynh5757** [comment]
Hi all, thanks for your patience. We have released `1.0.0` and are proud to say that this project is considered GA. Please update at your earliest convenience.

We're still working on improving the docs, please stay tuned for that!

**Comment by solidneeds** [comment]
> I was able to make a bit more progress.
> 
> For others that might be trying to figure out the managed Expo path here is my journey so far:
> 
> After looking at this: #143
> 
> I realized that this project itself is doing the pushkit registration, so that is not something I need to try to add in manually.
> 
> So, what I did:
> 
> * From my expo project, i added the @twilio/voice-react-native-sdk package through yarn.
> * Then, you need to create the 'custom dev client' through the expo, they have good docs on that but you run something like
>   `eas build --profile development --platform ios`
> * After that build works, you can install on your personal real device
> * After that client is downloaded to your device, you should see your app icon
> * The start the dev client server from your computer
>   `npx expo start --dev-client`
> * Opening up the app icon, I was able to run the app on my dev client, and make a call!
> 
> I need to setup the call handling and the incoming call notification still, but excited about being able to do an outbound call, without needed to 'eject' from expo
> 
> To all those building the project, thanks for the effort on this!

Could help me am trying run it on android using expo, what are the steps you followed? your help is appreciated

**Comment by dwbelliston** [comment]
> > I was able to make a bit more progress.
> > For others that might be trying to figure out the managed Expo path here is my journey so far:
> > After looking at this: #143
> > I realized that this project itself is doing the pushkit registration, so that is not something I need to try to add in manually.
> > So, what I did:
> > 
> > * From my expo project, i added the @twilio/voice-react-native-sdk package through yarn.
> > * Then, you need to create the 'custom dev client' through the expo, they have good docs on that but you run something like
> >   `eas build --profile development --platform ios`
> > * After that build works, you can install on your personal real device
> > * After that client is downloaded to your device, you should see your app icon
> > * The start the dev client server from your computer
> >   `npx expo start --dev-client`
> > * Opening up the app icon, I was able to run the app on my dev client, and make a call!
> > 
> > I need to setup the call handling and the incoming call notification still, but excited about being able to do an outbound call, without needed to 'eject' from expo
> > To all those building the project, thanks for the effort on this!
> 
> Could help me am trying run it on android using expo, what are the steps you followed? your help is appreciated

Hey, so this video from the expo team is really helpful for learning how to do a dev build:

https://www.youtube.com/watch?v=LUFHXsBcW6w

Should be a solid start for getting things running

**Comment by amitava82** [comment]
There is something missing. With those steps and dev client, it throws error on app start 

```Attempt to read from field 'com.twiliovoicereactnative.JSEventEmitter com.twiliovoicereactnative.VoiceApplicationProxy.jsEventEmitter' on a null object reference```

**Comment by solidneeds** [comment]
> There is something missing. With those steps and dev client, it throws error on app start
> 
> `Attempt to read from field 'com.twiliovoicereactnative.JSEventEmitter com.twiliovoicereactnative.VoiceApplicationProxy.jsEventEmitter' on a null object reference`

are you running on android, cause i get the same on android

**Comment by amitava82** [comment]
> > There is something missing. With those steps and dev client, it throws error on app start
> > `Attempt to read from field 'com.twiliovoicereactnative.JSEventEmitter com.twiliovoicereactnative.VoiceApplicationProxy.jsEventEmitter' on a null object reference`
> 
> are you running on android, cause i get the same on android

Yes

**Comment by amitava82** [comment]
That is for bare react native, not expo.

**Comment by solidneeds** [comment]
check this out https://docs.expo.dev/bare/installing-expo-modules/

maybe we have to manually edit android modules

**Comment by Victor-Sun** [comment]
I seem to be getting this issue as well. It was caused by the changes made from beta 3 to beta 4. Haven't found a way to fix the issue yet on React Native.

**Comment by ludvigbartholdsson** [comment]
Has anyone got incoming calls working with expo? Happy to pay for consultation.

**Comment by hueter** [comment]
Hi folks,

First off, this is a great library and very much needed for Twilio. 

I just wanted to mention that without official library support for Expo via a config plugin, a lot of us reading this are probably doing redundant work by rolling our own. Scanning this repo, I haven't found any published code for a working plugin so I'm about to try building one myself, but I will post it here and/or make a PR if I'm successful.

### For the maintainers

I think if you're a React Native library and you don't support Expo directly, your target audience is only a small minority of actual React Native users, since Expo is dominant in the RN space and is the recommendation from the actual RN docs!

The good news is, it's not a lot of effort to support Expo anymore. You just need to build a [config plugin](https://docs.expo.dev/config-plugins/introduction/), as others have mentioned. It's a small amount of JS to inject this already-functional React Native library's code into Expo's hidden `ios` and `android` folders. The reason it's hard _for us users_ is because it's literally all implementation details; I'm building this by reading your source code and trying to inject things one-by-one and follow the error messages.

### Possibly Helpful Resources

- [Here](https://github.com/expo/config-plugins) is a collection of config plugins written by the Expo team, that are functional examples of how to build them
  - In particular, here is one for [React Native CallKeep](https://github.com/expo/config-plugins/tree/main/packages/react-native-callkeep), a similar library.
- [react-native-permissions](https://github.com/zoontek/react-native-permissions) is an example of another library that maintains [their own first party Expo plugin](https://github.com/zoontek/react-native-permissions/blob/master/src/expo.ts) and simple instructions how to enable it.
- Expo Go (the platform app that Expo uses for standard workflows) will likely never support calling functionality due to its architectural limitations. Note that in order to test calling inbound and outbound (as others have mentioned), you'll have to do an on-device build. [You can do this locally](https://docs.expo.dev/guides/local-app-development/#local-app-compilation) without signing up for Expo or their cloud-based EAS build tool, the docs are just buried a bit.


I will try to update here and keep this thread going. Thanks!

**Comment by charliesantos** [comment]
I completely agree with you, @hueter! I understand that Expo support is definitely needed for this SDK, and it is currently on our roadmap. Please stay tuned, and let's keep this thread going!

**Comment by hueter** [comment]
Updating as promised!

### iOS works with a simple config plugin 🎉 

I was able to **get calling working for iOS**, inbound and outbound, using a simple Expo [config plugin](https://github.com/expo/config-plugins). As mentioned above, I based my code off the official [React Native Callkeep](https://github.com/expo/config-plugins/tree/main/packages/react-native-callkeep) support from Expo. [Here's a gist](https://gist.github.com/hueter/9b5a987c0f3e9f180dc46e5029829dfb) to demonstrate what worked for me.

### Android is difficult because of custom native code setup 🤔 

The Android setup is significantly more challenging, and I haven't figured it out yet. This is due to [the "Native code" docs](https://github.com/twilio/twilio-voice-react-native/blob/main/docs/getting-started.md#native-code):

> The native Android layer of the SDK exposes several helper classes that will need to be invoked in your existing native Android code.

For this, an Expo config plugin is not sufficient, and we need an [Expo Module](https://docs.expo.dev/modules/overview/).

Luckily, Expo modules expose [Android lifecycle listener callbacks](https://docs.expo.dev/modules/android-lifecycle-listeners/), which _should_ be enough to do the "native code" setup for this library. **In fact, this might make this library easier to install in Expo projects** because no native code editing would be necessary.

I wasn't able to create my own module, however, because I wasn't able to add `com.twiliovoicereactnative` as a dependency in `build.gradle`.  I don't think this library is published on the maven registry or whatever, and I couldn't find a way to include a dependency from `node_modules`. Because I need to do something like this:

<details>
  <summary>Click to expand!</summary>

```kt
package expo.modules.twiliovoicereactnative

import android.app.Application
import com.twiliovoicereactnative.VoiceApplicationProxy  // this line fails
import expo.modules.core.interfaces.ApplicationLifecycleListener

class TwilioVoiceReactNativeApplicationLifecycleListener : ApplicationLifecycleListener {
  private val voiceApplicationProxy: VoiceApplicationProxy

  override fun onCreate(application: Application) {
    voiceApplicationProxy = VoiceApplicationProxy(application)
    voiceApplicationProxy.onCreate();
    super.onCreate();
  }

  override fun onTerminate() {
    voiceApplicationProxy.onTerminate()
    super.onTerminate()
  }
}
```

</details>

I'm a bit out of my depth here with Android development, but I might have to fork this repo and make an Expo module within the source code here.


### Conclusion

It shouldn't be too difficult to support Expo given that iOS works as a config plugin and Android can tap into the lifecycle callbacks provided by Expo. I'm going to visit the Expo discord to see if there is a workaround to have a local dependency.

**Comment by Liam1809** [comment]
Any idea how we intergrate with android guys? thank you

**Comment by dwbelliston** [comment]
@hueter so you are thinking if this project was put on Maven, then the expo module path would be possible?

Looks like twilio has put projects on maven, so seems like it could happen:
https://mvnrepository.com/artifact/com.twilio.sdk/twilio-java-sdk

**Comment by hueter** [comment]
> @hueter so you are thinking if this project was put on Maven, then the expo module path would be possible?
> 
> Looks like twilio has put projects on maven, so seems like it could happen: https://mvnrepository.com/artifact/com.twilio.sdk/twilio-java-sdk

Possibly, if we cannot import the package already installed in `node_modules`. I don't know how to do that or if that's possible for Android.

But yeah we just need it so `import com.twiliovoicereactnative.VoiceApplicationProxy` works, then it's relatively trivial to build an Expo module I think.

**Comment by dwbelliston** [comment]
Okay, where you trying with a local expo module?

I dont know the android world well, so this may not be relevant.. but looking at the reference app here:

https://github.com/twilio/twilio-voice-react-native-app/blob/main/app/android/build.gradle#L32

Looks like there is an import from node_modules. Do you think that could be the 'import' we need in the modules `build.gradle` to get the `twilio react native` project installed?

**Comment by dwbelliston** [comment]
@hueter the module path seems like the right approach based on the docs page about the Android lifecycle listeners doc you shared above:

"""
The React Native module API does not provide any mechanism to hook into these, and so setup instructions for React Native libraries often include steps to copy code into these files. To simplify and automate setup and maintenance, the Expo Modules API provides a mechanism that allows your library to hook into Activity or Application functions.
"""

This seems very promising.

**Comment by dwbelliston** [comment]
Ive been trying to get a local expo module setup to do the setup mentioned here: https://github.com/twilio/twilio-voice-react-native/blob/main/docs/getting-started.md

I am running into same issue mentioned above by @hueter, the `com.twiliovoicereactnative` is not found when trying to import.

I tried many different configs with the `build.gradle` and `settings.gradle`... trying to get the package provided via the local node_modules. The code seemed to make sense, but just never got anywhere after trying for a while. I have never used gradle before, so could be a way off.

**Comment by dwbelliston** [comment]
I was able to do an `expo prebuild` to generate the android folder.. 

and then the files found here were helpful to reference to fit the `VoiceApplicationProxy` into the files.
https://github.com/twilio/twilio-voice-react-native/issues/348#issuecomment-2068747662

Im hoping ill be able to keep working down this path, of using prebuild at least to get it working. But some help and guidance on how to do the expo module would be a big win. Thanks! Hopefully these comments are helpful to others find a path forward too.

**Comment by dwbelliston** [comment]
@mhuynh5757 just checking if you saw the ask about putting this project into Maven, so we can explore working it into an expo module?

**Comment by periabyte** [comment]
IMO, Expo Support should already be Prioritized at this point, since it is already the **recommended** way in creating react native applications

> Using React Native frameworks, such as Expo, is now the recommended approach to create new apps

reference: https://reactnative.dev/blog/2024/06/25/use-a-framework-to-build-react-native-apps

**Comment by 0xbenon** [comment]
@dwbelliston and @hueter And, if I'm sure that my Android files will just only have this changes for react native, could be help create a config plugin with my native files modified, then in the config plugin, delete the default files and copy mines? Example, in your root expo app could we have?:

- kotlinModifyFiles/
     - MainActivity.kt
     - MainApplication.kt
     - MainReactNativeHost.kt

Then, replace the files in android/app/src/main/java/**/ ? Could this be possible or what do you think?

Thanks.

**Comment by 0xbenon** [comment]
> Hi [@dwbelliston](https://github.com/dwbelliston) and [@hueter](https://github.com/hueter) What about if I'm sure that my Android files will just only have this changes for react native, could be help create a config plugin with my kotlin files modified, then in the config plugin, delete the default files and copy mines? Example, in our root expo app dir could we have?:
> 
> * kotlinModifyFiles/
>   
>   * MainActivity.kt
>   * MainApplication.kt
>   * MainReactNativeHost.kt
> 
> Then, replace the files in android/app/src/main/java/**/ ? Could this be possible or what do you think?
> 
> Thanks.

Well, I tried it and it worked. For now I'll do it like this, while the twilio team adds the expo module.

expo-app-dir/configs/index.ts:

```ts
import { ConfigPlugin, withAndroidManifest } from "@expo/config-plugins";
import * as fs from "fs";
import * as path from "path";
import { copyKotlinFiles } from "./utils";

const withKotlinFiles: ConfigPlugin = (config) => {
  return withAndroidManifest(config, (config) => {
    const modifiedKotlinFilesPath = "kotlinModifiedFiles";
    
    const bundleIdentifierPath = config.modRawConfig.android?.package?.replaceAll(".", "/");
    const androidRootPath = config.modRequest.platformProjectRoot;
    const androidTargetPath = path.join(androidRootPath, `app/src/main/java/${bundleIdentifierPath}/`);
    
    copyKotlinFiles(modifiedKotlinFilesPath, androidTargetPath);
    
    return config;
  })
}

export default withKotlinFiles;
```

expo-app-dir/configs/utils.ts:

```ts
import * as fs from "fs";
import * as path from "path";

export const copyKotlinFiles = (
  sourcePath: string,
  targetPath: string
): void => {
  if (!fs.existsSync(sourcePath)) {
    throw new Error("The source path not exist.");
  }

  if (!fs.lstatSync(sourcePath).isDirectory())
    throw new Error("The source path is not a directory.");

  const files = fs.readdirSync(sourcePath);

  files.forEach(file => {
    console.log("Copying: ", file);
    copyFileSync(path.join(sourcePath, file), targetPath);
    console.log(" - File copied");
  });
};

const copyFileSync = (filePath: string, targetPath: string) => {
  let targetFile = targetPath;

  if (fs.existsSync(targetPath) && fs.lstatSync(targetPath).isDirectory()) {
     targetFile = path.join(targetPath, path.basename(filePath));
  }
  
  fs.writeFileSync(targetFile, fs.readFileSync(filePath));
};
```

**Comment by ludvigbartholdsson** [comment]
Also waiting for a solution for Expo

**Comment by gparkertwilio** [comment]
Hello all - to all Twilio customers following this issue who are interested in support for Expo with the Voice React Native SDK, may you please [reach out to our support team via a support ticket ](https://help.twilio.com/articles/360048500694-Contacting-Twilio-Support) at your next convenience in order for us to more formally track interest?  I'd also request that you please include your production account SID(s), use case, and expected traffic volumes as well.  Thank you!

**Comment by ludvigbartholdsson** [comment]
Don't forget to reach out !!

**Comment by mhuynh5757** [comment]
Hey all, please check out this Pinned Github Issue that we posted as a guide for getting Expo support working for your Expo apps: https://github.com/twilio/twilio-voice-react-native/issues/496

**Comment by tirth0** [comment]
Trying to follow #496 , I had a few doubts regarding the android implementation:

1. For the application lifecycle listener, would we need to create an implementation of VoiceApplicationProxy.VoiceReactNativeHost? Since VoiceApplicationProxy(...) expects an instance of reactNativeHost.

2. The given example for the ExpoModule uses ConnectOptions, CallRecords, CallListenerProxy data base from com.twiliovoicereactnative, however these functions seem to be package private `Cannot access 'CallListenerProxy': it is package-private in 'com. twiliovoicereactnative`. What's the way to implement this here?

Any help is appreciated, pretty new to this.

**Comment by mhuynh5757** [comment]
Hey @tirth0 thanks for reaching out and providing some feedback.

I've just updated #496 with some more info, specifically about forking the project for use. That might clear some things up.

You're correct in that the current constructor of the `VoiceApplicationProxy` needs a `ReactContext`. You can create a new constructor that accepts a `Context` instead of a `ReactContext` and use that as a private local variable in the `VoiceApplicationProxy`.

Since you'll be making a fork of this repo, those objects being package-private will be available to you.

**Comment by sagarshakya** [comment]
@mhuynh5757 The instructions in the Expo Support guide is very hard to follow for someone who doesn't have any native Android experience and hasn't worked with Expo Modules before.

**Comment by 0xbenon** [comment]
@mhuynh5757 Thanks for the guide.

It seems like it will take time for each team to create and maintain their own fork, since they need to migrate **TwilioVoiceReactNativeModule.java** to **ExpoModule.kt**. I think it would be better if the Twilio team provided an Expo‑compatible module out of the box, so that customers could simply install it. That way, individual customers wouldn’t have to spend time or acquire the in‑depth knowledge required to build and maintain a fork. In our case, we’ll continue using our current workaround (manually replacing native files for now) and may publish our own fork eventually—but it would be fantastic if Twilio could handle this upstream for everyone!

Thank you!

**Comment by mattrgibson** [comment]
> [@mhuynh5757](https://github.com/mhuynh5757) Thanks for the guide.
> 
> It seems like it will take time for each team to create and maintain their own fork, since they need to migrate **TwilioVoiceReactNativeModule.java** to **ExpoModule.kt**. I think it would be better if the Twilio team provided an Expo‑compatible module out of the box, so that customers could simply install it. That way, individual customers wouldn’t have to spend time or acquire the in‑depth knowledge required to build and maintain a fork. In our case, we’ll continue using our current workaround (manually replacing native files for now) and may publish our own fork eventually—but it would be fantastic if Twilio could handle this upstream for everyone!
> 
> Thank you!

It’s this exact reason that I ended up ditching Expo and starting my project from RN CLI. Seemed far easier to handle routing, navigation, bundling, etc., than to basically rewrite this repo from scratch in a custom Expo module.

**Comment by mhuynh5757** [comment]
Hey all, this is great feedback and I've forwarded it for my team to consider. However, since this issue is closed and very old, we'd like to ask about feedback regarding Expo to be made in a new thread. Thanks all.

**Comment by imthepitts** [comment]
Thanks, everyone, for your input on Expo over the past 2 years! We are collecting all the latest feedback for our future plans with the React Native SDK. 

Since we now have the [Expo guide published](https://github.com/twilio/twilio-voice-react-native/issues/496), we are going to lock this thread from further comment. If you have issues that come up while using the guide, please open a new issue so we can track and resolve them individually. 

We'd love to hear what you're building! Please feel free to reach out to me directly via email at zpitts @ Twilio.
